### DISCUSSION TITLE
Selenium not working after converting program to exe using pyinstaller

### DISCUSSION THREAD

**Discussion by Shivansh-yadav13** [pyinstaller/pyinstaller#7773] [Help] [answered]
After converting my program to exe program opens a new window of the application and does not start selenium.
I read solution provided in 
- #6777 

But using 
```
multiprocessing.freeze_support()
```
is still not working for me.

my code:

```
import math
from insta_api import Account
import multiprocessing


class Campaign:
    def __init__(
        self,
        comment_text,
        dm_text,
        ig_acc_list,
        lead_list,
        like_post=True,
        follow_lead=True,
    ):
        self.like_post = like_post
        self.follow_lead = follow_lead
        self.dm_text = dm_text
        self.comment_text = comment_text
        self.ig_acc_list = ig_acc_list
        self.lead_list = lead_list
        self.ig_api_acc_list = []
        self.assign_leads()

    def update_lead_list(self, new_lead_list):
        ...

    def assign_leads(self):
        ...

    def start_process(self):
        processes = []
        multiprocessing.freeze_support()
        for i, _ in enumerate(self.leads_sublist):
            p = multiprocessing.Process(target=self.ig_api_acc_list[i].start, args=[])
            p.start()
            processes.append(p)

        for p in processes:
            p.join()

```

**Comment by rokm** [comment]
You should call `multiprocessing.freeze_support()` once, at the start of your program (and before using any `multiprocessing` primitives in either your code or the code that you import).

But, was this a `multiprocessing` problem in the first place? If you create a test application that just calls `selenium` in main process, does it work when you freeze it using PyInstaller, or not?

**Reply by Shivansh-yadav13** [reply]
@rokm it works with normal .py file.

**Reply by Shivansh-yadav13** [reply]
It opens another application when the `start_process` is called.

**Reply by rokm** [reply]
> @rokm it works with normal .py file.

I imagine it does. But that does not answer the question of whether this is multiprocessing-related or not. 

That's why I'm asking you to try package a simple example that uses `selenium` without `multiprocessing`, and see if that works with PyInstaller or not (also because your example is not self-contained enough to try to reproduce the problem on my side).

**Reply by rokm** [reply]
> It opens another application when the start_process is called.

Another instance of *your* application?

**Reply by Shivansh-yadav13** [reply]
Yes

**Reply by rokm** [reply]
In that case, try moving the `freeze_support` call at the top of the file:

```
import multiprocessing
if __name__ == '__main__':
    multiprocessing.freeze_support()

import math
from insta_api import Account

class Campaign:
...
```

If that does not help, try adding
```
import sys
print(sys.argv)
```

at the very top of your program, to see what arguments the second instance is called with.

**Reply by Shivansh-yadav13** [reply]
> > @rokm it works with normal .py file.
> 
> I imagine it does. But that does not answer the question of whether this is multiprocessing-related or not.
> 
> That's why I'm asking you to try package a simple example that uses `selenium` without `multiprocessing`, and see if that works with PyInstaller or not (also because your example is not self-contained enough to try to reproduce the problem on my side).

@rokm My application works without `multiprocessing` packaged by pyinstaller

**Reply by rokm** [reply]
> > > @rokm it works with normal .py file.
> > 
> > 
> > I imagine it does. But that does not answer the question of whether this is multiprocessing-related or not.
> > That's why I'm asking you to try package a simple example that uses `selenium` without `multiprocessing`, and see if that works with PyInstaller or not (also because your example is not self-contained enough to try to reproduce the problem on my side).
> 
> @rokm My application works without `multiprocessing` packaged by pyinstaller

In that case, try moving freeze_support call to the top of your program (so it is executed before any 3rd party imports and the rest of your code), and if necessary, also add print(sys.argv) at the top of your program.

**Reply by Shivansh-yadav13** [reply]
```
import multiprocessing
import sys
if __name__ == '__main__':
    multiprocessing.freeze_support()
    print(sys.argv)
```

This is not working and nor printing anything in the console

**Reply by rokm** [reply]
What about:

```
import sys
print(sys.argv)

import multiprocessing
if __name__ == '__main__':
    multiprocessing.freeze_support()
```

**Comment by Shivansh-yadav13** [comment]
> What about:
> 
> ```
> import sys
> print(sys.argv)
> 
> import multiprocessing
> if __name__ == '__main__':
>     multiprocessing.freeze_support()
> ```

```
['D:\\learning_ctk\\dist\\monsterdm\\monsterdm.exe', '--multiprocessing-fork', 'parent_pid=29660', 'pipe_handle=3012']
```

this is the response, and then it pops another window of my app

**Reply by rokm** [reply]
Hmm... is the code executed in your entry-point script missing an `if __name__ == '__main__':` guard?

**Reply by rokm** [reply, marked as answer]
I.e., it should be like:

```python
import multiprocessing
if __name__ == '__main__':
    multiprocessing.freeze_support()

import math
from insta_api import Account

class Campaign:
  ...

if __name__ == '__main__':
   # Code that sets up your GUI, etc.
```

Without that `if __name__ == '__main__'` guard, the unguarded code will also end up being executed in any spawned `multiprocessing` process...

**Reply by Shivansh-yadav13** [reply]
@rokm 
MB I was not putting the code in my entry-point file.
now it's working, thanks.
